File: src/omnia/data/iterate/EmptyIterator.java
package omnia.data.iterate;

import java.util.Iterator;
import java.util.NoSuchElementException;
import java.util.Objects;
import java.util.function.Consumer;

public final class EmptyIterator<E> implements Iterator<E> {
  private static final EmptyIterator<?> EMPTY_ITERATOR = new EmptyIterator<>();

  public static <E> EmptyIterator<E> create() {
    @SuppressWarnings("unchecked")
    EmptyIterator<E> iterator = (EmptyIterator<E>) EMPTY_ITERATOR;
    return iterator;
  }

  private EmptyIterator() {}

  @Override
  public boolean hasNext() {
    return false;
  }

  @Override
  public E next() {
    throw new NoSuchElementException("iterator is empty");
  }

  @Override
  public void remove() {
    throw new IllegalStateException("iterator is empty");
  }

  @Override
  public void forEachRemaining(Consumer<? super E> action) {
    Objects.requireNonNull(action);
  }
}

File: src/omnia/data/iterate/EmptyIterator.kt
package omnia.data.iterate

import java.util.Objects
import java.util.function.Consumer

class EmptyIterator<out E> private constructor() : Iterator<E> {
  override fun hasNext(): Boolean {
    return false
  }

  override fun next(): E {
    throw NoSuchElementException("iterator is empty")
  }

  override fun forEachRemaining(action: Consumer<in E>) {
    Objects.requireNonNull(action)
  }

  companion object {
    private val EMPTY_ITERATOR: EmptyIterator<*> = EmptyIterator<Any>()

    @kotlin.jvm.JvmStatic
    fun <E> create(): EmptyIterator<E> {
      @Suppress("UNCHECKED_CAST")
      return EMPTY_ITERATOR as EmptyIterator<E>
    }
  }
}